### ISSUE TITLE
Missing type promotions when doing python API calls

### ISSUE THREAD

**Issue by Andreas-Krebbel** [numba/numba#9640]
I ran in to a problem with Numba on IBM Z (S/390). The issue potentially also affects other targets which require function arguments to be promoted (e.g. Mips).

The IBM Z ABI requires all integer types to be promoted to full register width by the caller. The callee will rely on this promotion to be done by the caller and will use an int32 input directly as 64 bit value in instructions without applying additional sign- or zeroextensions to it.

Right now Numba seems to violate that rule when it calls Python API routines as in:
```
    def tuple_new(self, count):
        fnty = ir.FunctionType(self.pyobj, [ir.IntType(32)])
        fn = self._get_function(fnty, name='PyTuple_New')
        return self.builder.call(fn, [self.context.get_constant(types.int32, count)])
```

I see this causing trouble in the testcase below, where a tuple with 2 elements is supposed to be allocated. Since there are some leftover bits in the upper 32 bits, PyTuple_New is invoked with some huge value and the allocation fails. A proper sign extension at the call-site would have prevented this. Experimentally changing the int32 types above to 64 bit makes the testcase succeed.

When LLVM generates IR, it consults the target definition via classifyArgumentType to figure out whether type promotion needs to be applied for function arguments and adds the "signext" or "zeroext" attributes to the IR of the function argument. From a quick look I couldn't find that information being provided by llvmlite yet. So I guess this would need to be the first step towards fixing this.

I'll try to have a look but would appreciate some guidance about what would be the right way to go about this.

Here the testcase I've extracted which triggers this problem on IBM Z:
```
import numpy as np
from numba import jit
from numba.typed import Dict

print("start")
@jit
def foo(k1, v1, k2):
    d = Dict()
    d[k1] = v1
    return d, d[k1]

k1 = (np.int32(1), np.int32(2))
k2 = 1
v1 = np.int32(123)

d, dk1 = foo(k1, v1, k2)
```

**Comment by sklam** [comment]
llvmlite do have `zeroext` and `signext`. Numba as the frontend will need to add the logic to insert those attributes when needed. It seems there's no harm just naively adding zeroext for all unsigned integers and signext for all signed integers.

**Comment by sklam** [comment]
Actually, for this case of `tuple_new`, it's just a bug. The function should take `Py_ssize_t` type which is `i64` on 64-bit platforms.

**Comment by sklam** [comment]
For the general LLVM-IR problem, we have been lucky that none of the supported platform cared. 

I wonder if this is only a problem when dealing functions written in other languages. If all of Numba doesn't emit `signext`/`zeroext` correctly, will all internal calls be consistent? Will passing a `i16` without proper extension work correctly all the time.

For all C FFI, is it only a problem when Numba jit code calls C? I presume it is the caller's responsibility to extend integers properly. If C calls Numba, will it matter that Numba doesn't set the proper integer extension on the argument attributes in LLVM?

**Comment by Andreas-Krebbel** [comment]
> llvmlite do have `zeroext` and `signext`. Numba as the frontend will need to add the logic to insert those attributes when needed.
Yes, what I couldn't find so far is the information from the target platform when to apply sign/zero extends on arguments. So basically a llvmlite API call for something like classifyArgumentType.

> It seems there's no harm just naively adding zeroext for all unsigned integers and signext for all signed integers.
Unfortunately it might trigger extra operations on platforms which actually don't need it.

**Comment by Andreas-Krebbel** [comment]
> For the general LLVM-IR problem, we have been lucky that none of the supported platform cared.
> 
> I wonder if this is only a problem when dealing functions written in other languages. If all of Numba doesn't emit `signext`/`zeroext` correctly, will all internal calls be consistent? Will passing a `i16` without proper extension work correctly all the time.

Yes, I think purely internal calls are ok. The calls would not adhere to the official ABI of the platform, but as long as it is consistent it should work.

> For all C FFI, is it only a problem when Numba jit code calls C? I presume it is the caller's responsibility to extend integers properly. If C calls Numba, will it matter that Numba doesn't set the proper integer extension on the argument attributes in LLVM?

As I understand it the signext/zeroext flags trigger different actions for the caller and the callee.

In the caller it imposes a requirement for the argument. So the caller has to do something extra to make it compatible.

In the callee it is some guarantee which can be exploited for further optimization. E.g. the callee might get rid of extra sign- or zeroextend operations which would be redundant.

So omitting the flags when compiling the callee should not be more than a missing optimization, while omitting the flags at call-site will likely cause a problem if the callee had the flags.  The pythonapi file seems to be used to let Numba compiled functions call into C compiled Python routines. So here currently we have the flags only in the callee.

**Comment by Andreas-Krebbel** [comment]
> Actually, for this case of `tuple_new`, it's just a bug. The function should take `Py_ssize_t` type which is `i64` on 64-bit platforms.

Indeed, good catch! I'll prepare a PR.

**Comment by MarkVeerasingam** [comment]
@gmarkall I think this issue can be closed
Resolved in PR https://github.com/numba/numba/pull/10396
